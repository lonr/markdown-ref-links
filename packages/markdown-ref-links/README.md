# `@lonr/markdown-ref-links`

> TODO: description

## Usage

```
const markdownRefLinks = require('@lonr/markdown-ref-links');

// TODO: DEMONSTRATE API
```
