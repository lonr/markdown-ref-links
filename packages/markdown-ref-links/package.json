{
  "name": "@lonr/markdown-ref-links",
  "version": "1.0.0",
  "description": "> TODO: description",
  "author": "lonr <lonr@live.cn>",
  "homepage": "",
  "license": "MIT",
  "main": "dist/markdown-ref-links.js",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "test": "echo \"Error: run tests from root\" && exit 1"
  },
  "dependencies": {
    "mdast-util-to-markdown": "^0.6.5",
    "unist-util-parents": "^1.0.3",
    "unist-util-visit": "^2.0.3"
  },
  "devDependencies": {
    "@types/jest": "^26.0.20",
    "jest": "^26.6.3",
    "remark": "^13.0.0",
    "rimraf": "^3.0.2",
    "rollup": "^2.40.0",
    "rollup-plugin-dts": "^3.0.0",
    "rollup-plugin-typescript2": "^0.30.0",
    "ts-jest": "^26.5.3",
    "ts-node": "^9.1.1",
    "tslib": "^2.1.0",
    "typescript": "^4.2.3"
  }
}
