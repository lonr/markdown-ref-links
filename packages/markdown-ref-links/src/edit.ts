// https://code.visualstudio.com/api/references/vscode-api#Position
// https://github.com/microsoft/vscode/blob/main/src/vs/editor/common/core/range.ts
// https://github.com/microsoft/vscode/blob/main/src/vs/workbench/api/common/extHostTypes.ts

export class Position {
  private _line: number;
  private _character: number;

  get line(): number {
    return this._line;
  }

  get character(): number {
    return this._character;
  }

  constructor(line: number, character: number) {
    this._line = line;
    this._character = character;
  }

  translate(lineDelta = 0, characterDelta = 0): Position {
    return new Position(this.line + lineDelta, this.character + characterDelta);
  }
}

export class Range {
  private _start: Position;
  private _end: Position;

  get start(): Position {
    return this._start;
  }

  get end(): Position {
    return this._end;
  }

  constructor(start: Position, end: Position) {
    this._start = new Position(start.line, start.character);
    this._end = new Position(end.line, end.character);
  }

  with(start: Position = this.start, end: Position = this.end): Range {
    return new Range(start, end);
  }
}

export class TextEdit {
  static replace(range: Range, newText: string): TextEdit {
    return new TextEdit(range, newText);
  }

  static insert(position: Position, newText: string): TextEdit {
    return TextEdit.replace(new Range(position, position), newText);
  }

  private _range: Range;
  private _newText: string | null;

  constructor(range: Range, newText: string | null) {
    this._range = range;
    this._newText = newText;
  }

  get newText(): string {
    return this._newText || '';
  }
  get range(): Range {
    return this._range;
  }
}
