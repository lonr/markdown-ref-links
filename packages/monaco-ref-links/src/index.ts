/*eslint-env browser*/
import * as monaco from 'monaco-editor/esm/vs/editor/editor.api';
import refLinks from '@lonr/remark-ref-links';
import remark from 'remark';

const inputModel = monaco.editor.createModel(
  '# Hello\n\n[hello](https://example.com)\n'
);
const outputModel = monaco.editor.createModel(
  remark().use(refLinks).processSync(inputModel.getValue()).toString()
);

const diffEditor = monaco.editor.createDiffEditor(
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  document.getElementById('container')!,
  {
    automaticLayout: true,
    originalEditable: true, // for left pane
    readOnly: true, // for right pane
  }
);

diffEditor.setModel({
  original: inputModel,
  modified: outputModel,
});

inputModel.onDidChangeContent(
  debounce(() => {
    const input = inputModel.getValue();
    const output = remark().use(refLinks).processSync(input).toString();
    outputModel.setValue(output);
  }, 1000)
);

function debounce(func: () => void, wait = 0) {
  let id: number | undefined;

  return () => {
    console.log();

    if (id !== undefined) {
      clearTimeout(id);
    }
    id = window.setTimeout(func, wait);
  };
}
