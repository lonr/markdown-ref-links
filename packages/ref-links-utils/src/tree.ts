import type { Heading, Link, Image, Definition } from 'mdast';
import type { Node } from 'unist';
import type { WithParent } from 'unist-util-parents';

export class Section {
  heading: WithParent<Heading>;
  introEnd: WithParent<Node>;
  images: WithParent<Image>[];
  links: WithParent<Link>[];
  children: Section[];

  constructor(heading: WithParent<Heading>) {
    this.heading = heading;
    this.introEnd = heading;
    this.images = [];
    this.links = [];
    this.children = [];
  }

  get lastChild(): Section | undefined {
    return this.hasChildren
      ? this.children[this.children.length - 1]
      : undefined;
  }

  get hasChildren(): boolean {
    return this.children.length > 0;
  }

  get outroEnd(): WithParent<Node> {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return this.hasChildren ? this.lastChild!.outroEnd : this.introEnd;
  }
}

export class Outline {
  maxDepth = 0;
  /// just to keep similar structure with Section
  introEnd: WithParent<Node>;
  images: WithParent<Image>[];
  links: WithParent<Link>[];
  definitions: Record<string, Record<string, Definition>>;
  existing: string[];
  children: Section[] = [];
  constructor(end: WithParent<Node>) {
    this.introEnd = end;
    this.images = [];
    this.links = [];
    this.definitions = {};
    this.existing = [];
  }
  get hasChildren(): boolean {
    return this.children.length > 0;
  }
  get lastChild(): Section | undefined {
    return this.hasChildren
      ? this.children[this.children.length - 1]
      : undefined;
  }
}
