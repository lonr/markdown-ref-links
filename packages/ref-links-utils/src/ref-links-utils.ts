import visit from 'unist-util-visit';
import { Outline, Section } from './tree';
export { Outline, Section };
import type { Root, Link, Image, Definition, Heading } from 'mdast';
import type { WithParent } from 'unist-util-parents';

export function genOutline(root: WithParent<Root>): Outline {
  const outline = new Outline(root);
  let lastSection: Section | Outline = outline;

  for (const child of root.children) {
    if (child.type === 'heading') {
      if (child.depth > outline.maxDepth) {
        outline.maxDepth = child.depth;
      }
      let p: Outline | Section = outline;
      for (let depth = 0; depth < child.depth; depth += 1) {
        if (p.children.length === 0) {
          break;
        } else {
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          const lastChild: Section = p.lastChild!;
          if (lastChild.heading.depth < child.depth) {
            p = lastChild;
          }
        }
      }
      lastSection = new Section(child as WithParent<Heading>);
      p.children.push(lastSection);
    } else {
      lastSection.introEnd = child;
    }

    visit(
      child,
      ['image', 'link'],
      (node: WithParent<Image> | WithParent<Link>) => {
        if (node.type === 'link') {
          lastSection.links.push(node);
        } else if (node.type === 'image') {
          lastSection.images.push(node);
        }
      }
    );
  }

  // https://github.com/remarkjs/remark-reference-links/blob/f14a49794590a3d3e97f65deabd82ad5e8725dc3/index.js#L22
  visit(root, 'definition', (definition: WithParent<Definition>) => {
    const url = definition.url;
    const definitions = outline.definitions;
    outline.existing.push(definition.identifier);
    if (!(url in definitions)) {
      definitions[url] = {};
    }
    definitions[url][definition.title ?? ''] = definition;
  });

  return outline;
}
