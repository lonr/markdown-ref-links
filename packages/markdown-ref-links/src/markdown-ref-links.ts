/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { genOutline, Outline, Section } from '@lonr/ref-links-utils';
import parents, { WithParent } from 'unist-util-parents';
import unified from 'unified';
import type { Node } from 'unist';
import type {
  Root,
  Heading,
  Link,
  Image,
  Definition,
  LinkReference,
  ImageReference,
} from 'mdast';
import remarkParse from 'remark-parse';
import toMarkdown from 'mdast-util-to-markdown';
import { Position, Range, TextEdit } from './edit';

export function format(
  markdown: string,
  userSettings?: { maxDepth?: number; allowDuplicate?: boolean }
): string {
  const settings = {
    ...{ maxDepth: -1, allDuplicate: false },
    ...userSettings,
  };
  const processor = unified().use(remarkParse);
  const root = parents(processor.parse(markdown) as Root);
  const outline = genOutline(root);
  const { definitions, existing } = outline;
  const edits: TextEdit[] = [];
  let id = 0;
  settings.maxDepth =
    settings.maxDepth >= 0
      ? settings.maxDepth
      : outline.maxDepth + settings.maxDepth + 1;

  const topInsertAnchor = root.children[root.children.length - 1];
  // handles Markdown with no headers or multiple h1 headers
  const outlineInsertAnchor =
    outline.hasChildren && settings.maxDepth !== 0
      ? outline.introEnd
      : topInsertAnchor;

  for (const linkOrImage of [...outline.links, ...outline.images]) {
    // if root has no children, it should have no links too
    convert(linkOrImage, outlineInsertAnchor);
  }

  const sectionsInsertAnchor =
    settings.maxDepth === 0 ? outlineInsertAnchor : undefined;

  for (const section of outline.children) {
    traverseSection(section, sectionsInsertAnchor);
  }

  return markdown;

  function traverseSection(
    section: Section,
    insertAnchor?: WithParent<Node>
  ): void {
    const links = section.links;
    const images = section.images;
    const underMaxDepth = section.heading.depth < settings.maxDepth;
    const sectionInsertAnchor =
      insertAnchor ?? (underMaxDepth ? section.introEnd : section.outroEnd);

    for (const linkOrImage of [...links, ...images]) {
      convert(linkOrImage, sectionInsertAnchor);
    }

    const childrenInsertAnchor =
      insertAnchor ?? (underMaxDepth ? undefined : section.outroEnd);

    for (const child of section.children) {
      traverseSection(child, childrenInsertAnchor);
    }
  }

  function convert(
    linkOrImage: WithParent<Link | Image>,
    insertAnchor: WithParent<Node>
  ) {
    // eslint-disable-next-line prefer-const
    let { type, url, title } = linkOrImage;
    title ??= '';
    let titles: Record<string, Definition> = {};
    let identifier: string;

    let definitionExisted = false;

    if (url in definitions) {
      titles = definitions[url];
    } else {
      definitions[url] = titles;
    }

    if (title in titles) {
      identifier = titles[title].identifier;
      definitionExisted = true;
    } else {
      do {
        identifier = `${++id}`;
      } while (existing.includes(identifier));
    }

    const definition: Definition = {
      type: 'definition',
      identifier: identifier,
      title,
      url: url,
    };

    titles[title] = definition;
    if (!definitionExisted || settings.allDuplicate) {
      insertAfter(insertAnchor, definition);
    }

    const replacement: LinkReference | ImageReference = {
      type: `${type}Reference` as const,
      identifier: identifier,
      referenceType: 'full',
    } as LinkReference | ImageReference;

    if (type === 'link') {
      replacement.children = linkOrImage.node.children;
    } else if (type === 'image') {
      replacement.alt = linkOrImage.node.alt;
    }
    replace(linkOrImage, replacement);
  }

  function replace(node: WithParent<Node>, newNode: Node) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    let { line, column } = node.position!.start;
    const start = new Position(line - 1, column - 1);
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    ({ line, column } = node.position!.end);
    const end = new Position(line - 1, column - 1);
    const range = new Range(start, end);
    const newText = toMarkdown(newNode);
    edits.push(TextEdit.replace(range, newText))
  }

  function insertAfter(node: WithParent<Node>, newNode: Node) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const { line } = node.position!.start;
    // const position = new Position(line - 1 + 1, 0);
    const position = new Position(line, 0);
    const newText = toMarkdown(newNode);
    edits.push(TextEdit.insert(position, newText));
  }

}
