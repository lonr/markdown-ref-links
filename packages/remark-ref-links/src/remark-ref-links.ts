// eslint is broken on these two rules?
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
// (The MIT License)
// Copyright (c) 2015 Titus Wormer <tituswormer@gmail.com>
// https://github.com/remarkjs/remark-reference-links/blob/main/index.js


// import { genOutline, Section } from '../../ref-links-utils/src/ref-links-utils';
import { genOutline, Section } from '@lonr/ref-links-utils';
import parents, { WithParent } from 'unist-util-parents';
import type { Plugin, Settings } from 'unified';
import type { Node } from 'unist';
import type {
  Root,
  Definition,
  LinkReference,
  ImageReference,
  Image,
  Link,
} from 'mdast';

const refLinks: Plugin = function attacher(userSettings) {
  const settings = {
    ...{ maxDepth: -1, allDuplicate: false },
    ...userSettings,
  };

  return function transformer(tree, file) {
    let id = 0;
    const root = parents(tree as Root);
    const outline = genOutline(root);
    const { definitions, existing } = outline;
    settings.maxDepth =
      settings.maxDepth >= 0
        ? settings.maxDepth
        : outline.maxDepth + settings.maxDepth + 1;

    const topInsertAnchor = root.children[root.children.length - 1];
    // handles Markdown with no headers or multiple h1 headers
    const outlineInsertAnchor =
      outline.hasChildren && settings.maxDepth !== 0
        ? outline.introEnd
        : topInsertAnchor;

    for (const linkOrImage of [...outline.links, ...outline.images]) {
      // if root has no children, it should have no links too
      convert(linkOrImage, outlineInsertAnchor);
    }

    const sectionsInsertAnchor =
      settings.maxDepth === 0 ? outlineInsertAnchor : undefined;

    for (const section of outline.children) {
      traverseSection(section, sectionsInsertAnchor);
    }

    /**
     * traverse sections and replace links and images
     * @param {Section} section
     * @param {WithParent<Node>} insertAnchor used if a section need to insert defs after it's ancestor's outroEnd
     */
    function traverseSection(
      section: Section,
      insertAnchor?: WithParent<Node>
    ): void {
      const links = section.links;
      const images = section.images;
      const underMaxDepth = section.heading.depth < settings.maxDepth;
      const sectionInsertAnchor =
        insertAnchor ?? (underMaxDepth ? section.introEnd : section.outroEnd);

      for (const linkOrImage of [...links, ...images]) {
        convert(linkOrImage, sectionInsertAnchor);
      }

      const childrenInsertAnchor =
        insertAnchor ?? (underMaxDepth ? undefined : section.outroEnd);

      for (const child of section.children) {
        traverseSection(child, childrenInsertAnchor);
      }
    }

    function convert(
      linkOrImage: WithParent<Link | Image>,
      insertAnchor: WithParent<Node>
    ) {
      // eslint-disable-next-line prefer-const
      let { type, url, title } = linkOrImage;
      title ??= '';
      let titles: Record<string, Definition> = {};
      let identifier: string;

      let definitionExisted = false;

      if (url in definitions) {
        titles = definitions[url];
      } else {
        definitions[url] = titles;
      }

      if (title in titles) {
        identifier = titles[title].identifier;
        definitionExisted = true;
      } else {
        do {
          identifier = `${++id}`;
        } while (existing.includes(identifier));
      }

      const definition: Definition = {
        type: 'definition',
        identifier: identifier,
        title,
        url: url,
      };

      titles[title] = definition;
      if (!definitionExisted || settings.allDuplicate) {
        insertAfter(insertAnchor, definition);
      }

      const replacement: LinkReference | ImageReference = {
        type: `${type}Reference` as const,
        identifier: identifier,
        referenceType: 'full',
      } as LinkReference | ImageReference;

      if (type === 'link') {
        replacement.children = linkOrImage.node.children;
      } else if (type === 'image') {
        replacement.alt = linkOrImage.node.alt;
      }
      replace(linkOrImage, replacement);
    }

    function replace(node: WithParent<Node>, newNode: Node) {
      // node is at least a child of root
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const parent = node.parent!;
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const index = node.parent!.children.indexOf(node);
      parent.node.children[index] = newNode;
    }

    function insertAfter(node: WithParent<Node>, newNode: Node) {
      // root!
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const parent = node.parent!;
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      let index = node.parent!.children.indexOf(node);
      // fix the inserting order
      do {
        index += 1;
      } while (
        index < parent.node.children.length &&
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        parent.node.children[index]!.type !== 'heading'
      );
      parent.node.children.splice(index, 0, newNode);
    }
  };
};

export default refLinks;
