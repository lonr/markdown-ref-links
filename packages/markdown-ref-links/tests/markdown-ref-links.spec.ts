import remark from 'remark';
import { format } from '../src/markdown-ref-links';
import fs from 'fs';
import { join, extname } from 'path';

test('fixtures with default settings', () => {
  const root = join(__dirname, 'fixtures');
  // https://github.com/remarkjs/remark-reference-links/blob/main/test.js
  fs.readdirSync(root)
    .filter((name) => extname(name) === '.md')
    .forEach((name) => {
      const input = fs.readFileSync(join(root, name));
      const actual = format(input.toString());
      expect(actual).toMatchSnapshot();
    });
});
