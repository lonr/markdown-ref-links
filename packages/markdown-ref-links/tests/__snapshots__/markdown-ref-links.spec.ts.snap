// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixtures with default settings 1`] = `
"# h1

p1

[linkA](http://a.com \\"title a\\") 

## h2

p1.1

[linkADup](http://a.com \\"title a\\")

[linkB](http://b.com)

### h3

p1.1.1

[linkC](https://c.com)

[2]: https://c.com

### h3dup

[linkD](https://d.com)

## h2

p1.2

![image1](https://foo.com/image)
"
`;

exports[`fixtures with default settings 2`] = `
"big bang

[link0](http://example.com \\"Example Domain\\") 

## h1

p1

[linkA](http://a.com \\"title a\\") 

# h

p2

[linkADup](http://a.com \\"title a\\")

[linkB](http://b.com)

# h3

p3

[linkC](https://c.com)

[2]: https://c.com

## h2

p3.1

[linkD](https://d.com)

![image1](https://foo.com/image)

###### h6
"
`;
